File: StaxLang.JS/rational.ts
import { abs, gcd } from './integer';
import { floatify } from './types';

export class Rational {
    public numerator: bigint;
    public denominator: bigint;

    public valueOf(): number {
        return Number(Number(this.numerator) / Number(this.denominator));
    }

    public toString(): string {
        return `${this.numerator}/${this.denominator}`;
    }

    constructor(num: bigint, den: bigint) {
        this.numerator = num;
        this.denominator = den;
        this.reduce();
    }

    invert() {
        return new Rational(this.denominator, this.numerator);
    }

    add(other: Rational): Rational {
        return new Rational(
            this.numerator * other.denominator + other.numerator * this.denominator,
            this.denominator * other.denominator);
    }

    subtract(other: Rational) {
        return new Rational(
            this.numerator * other.denominator - other.numerator * this.denominator,
            this.denominator * other.denominator);
    }

    multiply(other: Rational): Rational {
        return new Rational (
            this.numerator * other.numerator,
            this.denominator * other.denominator);
    }

    divide(other: Rational) {
        return new Rational (
            this.numerator * other.denominator,
            this.denominator * other.numerator);
    }

    negate() {
        return new Rational(-this.numerator, this.denominator);
    }

    abs() {
        return new Rational(abs(this.numerator), abs(this.denominator));
    }

    floor() {
        if (this.numerator >= 0n) return this.numerator / this.denominator;
        return (this.numerator - this.denominator + 1n) / this.denominator;
    }

    ceiling() {
        return -this.negate().floor();
    }

    mod(other: Rational) {
        other = other.abs();
        let intPart = this.divide(other).floor();
        return this.subtract(other.multiply(new Rational(intPart, 1n)));
    }

    equals(other: Rational) {
        return this.numerator === other.numerator && this.denominator === other.denominator;
    }

    private reduce() {
        if (this.denominator === 0n) throw new Error("rational divide by zero");
        let divisor = gcd(this.numerator, this.denominator);
        this.numerator /= divisor;
        this.denominator /= divisor;
        if (this.denominator < 0n) {
            this.numerator = -this.numerator;
            this.denominator = -this.denominator;
        }
    }
}

export function rationalize(arg: number): Rational {
    const significantBits = 50, epsilon = arg / 2 ** significantBits;;

    if (arg < 0) return rationalize(-arg).negate();
    if (arg % 1 == 0) return new Rational(BigInt(arg), 1n);

    type rat = [bigint, bigint];
    let
        left: rat = [BigInt(Math.floor(arg)), 1n],
        right: rat = [BigInt(Math.ceil(arg)), 1n],
        best: rat = [0n, 1n], mediant: rat;
    let bestError = Number.POSITIVE_INFINITY, lastMove = 1n;

    do { // Stern-Brocot binary search
        mediant = lastMove < 0n
            ? [right[0] - lastMove * left[0], right[1] - lastMove * left[1]]
            : [left[0] + lastMove * right[0], left[1] + lastMove * right[1]];

        let error = floatify(mediant[0]) / floatify(mediant[1]) - arg;
        if (isNaN(error)) error = 0;

        if (error > 0) {
            if (lastMove < 0n) {
                right = mediant;
                lastMove *= 2n;
            }
            else {
                lastMove /= 2n;
                if (lastMove === 0n) lastMove = -1n;
            }
        }
        else {
            if (lastMove > 0n) {
                left = mediant;
                lastMove *= 2n;
            }
            else {
                lastMove /= 2n;
                if (lastMove === 0n) lastMove = 1n;
            }
        }
        if (Math.abs(error) < bestError) [best, bestError] = [mediant, Math.abs(error)];
    } while (bestError > epsilon);
    return new Rational(best[0], best[1]);
}

export const zero = new Rational(0n, 1n);
export const one = new Rational(1n, 1n);

File: StaxLang.Interpreter/Rational.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace StaxLang {
    public class Rational : IComparable {
        public BigInteger Num { get; private set; }
        public BigInteger Den { get; private set; }

        public Rational(BigInteger num, BigInteger den) {
            Num = num;
            Den = den;
            Reduce();
        }

        private void Reduce() {
            var reduction = BigInteger.GreatestCommonDivisor(Num, Den);
            Num /= reduction;
            Den /= reduction;

            if (Den < 0) {
                Num *= -1;
                Den *= -1;
            }
        }

        public Rational AbsoluteValue() => new Rational(BigInteger.Abs(Num), BigInteger.Abs(Den));

        public static Rational operator -(Rational a, Rational b) => new Rational(a.Num * b.Den - b.Num * a.Den, a.Den * b.Den);
        public static Rational operator +(Rational a, Rational b) => new Rational(a.Num * b.Den + b.Num * a.Den, a.Den * b.Den);
        public static Rational operator /(Rational a, Rational b)=> new Rational(a.Num * b.Den, a.Den * b.Num);
        public static Rational operator *(Rational a, Rational b)=> new Rational(a.Num * b.Num, a.Den * b.Den);
        public static Rational operator %(Rational a, Rational b) => a - (a / b).Floor() * b;
        public static Rational operator -(Rational a) => a * -1;
        public static bool operator ==(Rational a, Rational b) => a.Equals(b);
        public static bool operator !=(Rational a, Rational b) => !a.Equals(b);
        public static bool operator <(Rational a, Rational b) => a.Num * b.Den < b.Num * a.Den;
        public static bool operator >(Rational a, Rational b) => a.Num * b.Den > b.Num * a.Den;
        public static bool operator <=(Rational a, Rational b) => a.Num * b.Den <= b.Num * a.Den;
        public static bool operator >=(Rational a, Rational b) => a.Num * b.Den >= b.Num * a.Den;

        public static implicit operator Rational(int n) => new Rational(n, 1);
        public static implicit operator Rational(BigInteger n) => new Rational(n, 1);
        public static explicit operator double(Rational r) => (double)r.Num / (double)r.Den;

        public BigInteger Floor() {
            if (Num < 0) return (Num - Den + 1) / Den; 
            else return Num / Den;
        }

        public BigInteger Ceil() {
            if (Num < 0) {
                return Num / Den; 
            }
            else {
                return (Num + Den - 1) / Den;
            }
        }

        public override string ToString() => $"{Num}/{Den}";
        public string ToString(IFormatProvider format) => Num.ToString(format) + "/" + Den.ToString(format);
        public override bool Equals(object obj) => obj is Rational r && Num == r.Num && Den == r.Den;
        public override int GetHashCode() => Num.GetHashCode() ^ (Den.GetHashCode() * 37);

        public int CompareTo(object obj) {
            var r = (Rational)obj;
            if (this > r) return 1;
            if (this < r) return -1;
            if (this == r) return 0;
            throw new Exception("Rational compare sanity failed");
        }

        public static Rational Rationalize(double arg) {
            const int SignificantBits = 50;

            if (arg < 0) return -Rationalize(-arg);
            if (arg % 1 == 0) return new Rational(new BigInteger(arg), 1);

            double epsilon = arg / Math.Pow(2, SignificantBits), bestError = double.PositiveInfinity;
            (BigInteger Num, BigInteger Den)
                left = (new BigInteger(Math.Floor(arg)), 1),
                right = (new BigInteger(Math.Ceiling(arg)), 1),
                best = default, mediant;
            BigInteger lastMove = 1;

            do { // Stern-Brocot binary search
                mediant = lastMove < 0
                    ? (-lastMove * left.Num + right.Num, -lastMove * left.Den + right.Den)
                    : (left.Num + lastMove * right.Num, left.Den + lastMove * right.Den);

                double error = (double)mediant.Num / (double)mediant.Den - arg;
                if (double.IsNaN(error)) error = 0;

                if (error > 0) {
                    if (lastMove < 0) {
                        right = mediant;
                        lastMove *= 2;
                    }
                    else if ((lastMove /= 2) == 0) lastMove = -1;
                }
                else {
                    if (lastMove > 0) {
                        left = mediant;
                        lastMove *= 2;
                    }
                    else if ((lastMove /= 2) == 0) lastMove = 1;
                }
                if (Math.Abs(error) < bestError) (best, bestError) = (mediant, Math.Abs(error));
            } while (bestError > epsilon);
            return new Rational(best.Num, best.Den);
        }
    }
}
